### ISSUE TITLE
Handle RTC through hard reset in Gambatte core

### ISSUE THREAD

**Issue by TiKevin83** [TASVideos/BizHawk#1895]
### Summary
MrWint's update which added cycle accuracy to RTC didn't handle the corner case of hard resets during games that support RTC

### Repro
TAS Pokemon Crystal and hard reset sometime after setting the time and saving

### Output
The core behaves non-determinstically after hard resetting, tracking RTC from the system clock instead of the cycles of the RTC on the cartridge.

This should be fixable somewhat trivially by both confirming that the new cycle-accurate cart RTC time is appended to the end of .SaveRAM and getting BizHawk to continue using that value through hard reset instead of using the system clock

**Comment by alyosha-tas** [comment]
This should be handled internally by the core, Can the RTC simply not be effected by hard reset?

**Comment by alyosha-tas** [comment]
I think the best solution here is to add a bool here:

`void gambatte::setInitState(SaveState &state, const bool cgb)`

to check if the RTC should be reset or not. BizHawk hard reset calls this upon reset the same as it would upon first loading the core. Which in turn calls the system clock function to set the RTC bas time.

I'm actually not a fan of using the system clock at all, but it works I guess.

**Comment by TiKevin83** [comment]
@alyosha-tas I looked at this one briefly and wasn't understanding what was going on at all, are you able to take a look or expand on your previous note?

**Comment by alyosha-tas** [comment]
I tried (and failed) to fix this about a month ago. I'll give it a fresh look (and leave notes this time.)

**Comment by alyosha-tas** [comment]
`	if (reset_RTC)
	{
		state.time.seconds = 0;
	}`

Making this deterministic seems to be just a matter of doing something similar to the above in initstate.cpp (after adding the bool condition reset_RTC.) However, this is technically not correct as it does not account for the 8 cycles of CPU time that should be taken for the cpu to power back up. In reality this doesn't really matter since the hard reset is not instant anyway, so missing a couple cycles is irrelevant compared to the un-quantifiable time it takes to toggle power.

What really should happen is a reset operation that takes 8 cycles should be added to the cpu so we can do away with the constantly starting from cycle 8 instead of 0 initial state stuff. (this is what GBHawk does already, it turned out much cleaner this way.)

**Comment by CasualPokePlayer** [comment]
I took a stab at this issue and found some potential issues:

The initial state is somewhat different going from upstream to bizhawk. Upstream separates the "time" variables and the "FF the save" thing into a separate init cart state function (and only calls it on rom load, reset just uses init state), and for init state does something different with lastcycles. Potentially the issue?

As another note, the save state code for time looks very wrong. Upstream calls the timeFromCycles or cyclesFromTime function before save stating, while that doesn't seem to happen for bizhawk. From experience, I haven't experience any desyncs TASing Pokemon Gen 2, so this might not be needed/already handled somewhere and I don't see it, but it looks like a potential issue (might not be related to this? I don't know).

Should also note that I have experienced with upstream _very_ wrong behavior in the case the timestamp loaded ends up being a later time than the actual current system time (resulting in some functions calculating a "negative" (or really just really large positive) number in some variables). That could maybe just be the route cause? Where this is actually occurring (if it even is) however I can't figure out.

**Comment by CasualPokePlayer** [comment]
So I managed to fix RTC. Well, it was somewhat complicated.

There was a lot of difference between upstream and Bizhawk. Bizhawk seemed to just placed a bunch of uinit32_t instead of time_t, so I changed that. Upstream also just got rid of its own reliance with loadSaveData for resetting, so I changed that too. Upstream also used gettimeofday while Bizhawk (due to being built with MSVC rather than with a Unix-like build system) used a workaround with high_resolution_clock. Upon further inspection, it seems like high_resolution_clock was implementation dependent, and so happens MSVC alias it to steady_clock, which is bad when you want to track time between rom being unloaded and loaded again (casual play only, irrelevant for movies, which won't be using this now anyways). So I just changed it to system_clock (which is what would be wanted here). However, loadSaveData was still causing issues in the case the core was restarted/rom was reloaded, and in principle I can't avoid using loadSaveData. Which lead to the core cause of this issue:

The data was being grabbed for RTC but wasn't &'d by 0xFF at each step so that caused issues lmao. Adding the & 0xFF fixed this issue, so I can probably say RTC is completely fixed now.
